Xml doc bugfixes

BillWagner opened the pull request:
Fixes three issues:

Fixes #771. The string IDs for methods that are explicit implementations of members of generic interfaces have the `<` and `>` characters replaced with `{` and `}`.
Fixes #772:  single dimension arrays don't specify the lower bound in their string ids. Instead of `[0]` its `[]`.
Fixes #915:  The example didn't include the "name=" attributes on parameters.

The fixes are in three distinct commits to make it easier to review.

BillWagner committed:
Correct ID string for generic elements

Fixes #771

Add instances where the `<` and `>` characters are replaced by `{` and `}` in the string IDs.

BillWagner committed:
Fix array encoding

Fixes #772

Single dimension arrays are encoded as "[]" rather than "[0]".

BillWagner committed:
Add name attribute to param tags

Fixes #915

In the example in D5, add the "name=" attribute on all param tags.

jskeet commented on standard/documentation-comments.md:
```diff
@@ -1031,8 +1031,8 @@ namespace Graphics
         /// </summary>
-        /// <param><c>xPosition</c> is the new Point's x-coordinate.</param>
-        /// <param><c>yPosition</c> is the new Point's y-coordinate.</param>
+        /// <param name="xPosition"><c>xPosition</c> is the new Point's x-coordinate.</param>
```
At some point we should improve the docs to remove the "xPosition is..." bits. But not now :)

BillWagner replied:
I had wanted to do it anyway. Thanks for pointing that out.

BillWagner replied:
I went ahead and just did it.

BillWagner committed:
Address nit

I thought about removing the `<c>` tags in my previous commit, but I was waiting for other opinions, and wanted to verify that there was at least one tag like that still in the example.

jskeet reviewed (approved):
Hooray, thanks :)

KalleOlaviNiemitalo commented:
Now I wonder how the name of a type is formatted in these if it involves an `extern alias`.  Particularly if class C has conversion operators to both X::T and Y::T, where X and Y are extern aliases, and T is the same identifier in both but means different types.

KalleOlaviNiemitalo commented:
The extern alias thing is a bit related to https://github.com/dotnet/docfx/issues/2620

jskeet commented on standard/documentation-comments.md:
```diff
@@ -691,12 +691,12 @@ The documentation generator observes the following rules when it generates the I
     - Arguments that represent generic types have an appended “`'`” character followed by the number of type parameters
     - Arguments having the `in`, `out` or `ref` modifier have an `@` following their type name. Arguments passed by value or via `params` have no special notation.
-    - Arguments that are arrays are represented as `[` *lowerbound* `:` *size* `,` … `,` *lowerbound* `:` *size* `]` where the number of commas is the rank less one, and the lower bounds and size of each dimension, if known, are represented in decimal. If a lower bound or size is not specified, it is omitted. If the lower bound and size for a particular dimension are omitted, the “`:`” is omitted ...
+    - Arguments that are multi-dimensional arrays are represented as `[` *lowerbound* `:` *size* `,` … `,` *lowerbound* `:` *size* `]` where the number of commas is the rank less one, and the lower bounds and size of each dimension, if known, are represented in decimal. If a lower bound or size is not specified, it is omitted. If the lower bound and size for a particular dimension are omitted, th ...
```
```suggestion
    - Arguments that are arrays are represented as `[` *lowerbound* `:` *size* `,` … `,` *lowerbound* `:` *size* `]` where the number of commas is the rank less one, and the lower bounds and size of each dimension, if known, are represented in decimal. If a lower bound or size is not specified, it is omitted. If the lower bound and size for a particular dimension are omitted, the “`:`” is omitted as well. Jagged arrays are represented by one “`[]`” per level. Single dimensional arrays omit the lower bound when the lower bound is 0 (the default) (§17.1).
```

gafter committed:
Update standard/documentation-comments.md

Co-authored-by: Jon Skeet <skeet@pobox.com>

jskeet commented:
> Now I wonder how the name of a type is formatted in these if it involves an `extern alias`.

Meeting note: We're not sure either. Please feel free to experiment and either raise a PR or an issue with your findings.
